File: HuddleCore/Sources/HuddleCore/RoomState.swift
/// The lifecycle of a room. Every state has a guaranteed exit — no dead ends.
///
/// Locked flow: `LOBBY → ACTIVE → TALLY → REVEALING → MATCHED | NO_RESULT`.
/// The host may end the session at any time (modeled as a transition to
/// `.noResult`, the graceful terminal).
///
/// Raw values are SCREAMING_CASE because they are the wire format: the
/// backend will emit these exact strings, and the client must decode them
/// byte-for-byte. Do not rename casually — that is a protocol change.
public enum RoomState: String, Codable, Sendable, CaseIterable {
    case lobby = "LOBBY"
    case active = "ACTIVE"
    case tally = "TALLY"
    case revealing = "REVEALING"
    case matched = "MATCHED"
    case noResult = "NO_RESULT"

    /// Terminal states have no exits. A room that reaches one is resolved.
    public var isTerminal: Bool {
        switch self {
        case .matched, .noResult: true
        case .lobby, .active, .tally, .revealing: false
        }
    }

    /// Swipes are only accepted while ACTIVE. In particular REVEALING is
    /// input-closed: the server directs the reveal pacing and any swipe
    /// arriving in that window must be rejected, not queued.
    public var acceptsSwipes: Bool {
        self == .active
    }

    /// The full transition table. Keeping it in one place (rather than
    /// scattered `if` checks) makes the state machine auditable: the tests
    /// walk this table exhaustively.
    public var allowedTransitions: Set<RoomState> {
        switch self {
        case .lobby:
            // Start the round, or host ends before anything happened.
            [.active, .noResult]
        case .active:
            // All present finished swiping, or the inactivity timeout fired.
            // Either trigger lands in TALLY; host may still end.
            [.tally, .noResult]
        case .tally:
            // .revealing — a result (match, tie resolution, blind pick) is
            //   ready to be presented.
            // .active — overtime: eliminate-lowest re-run or a fresh-deck
            //   round after zero valid yes. Hard-capped by the engine (the
            //   cap lives with the round logic, not in this table).
            [.revealing, .active, .noResult]
        case .revealing:
            [.matched, .noResult]
        case .matched, .noResult:
            []
        }
    }

    public func canTransition(to next: RoomState) -> Bool {
        allowedTransitions.contains(next)
    }
}

File: backend/src/domain/roomState.ts
/**
 * MIRROR of HuddleCore/Sources/HuddleCore/RoomState.swift — the TS half
 * of the wire contract. The strings below ARE the protocol; they must
 * match the Swift raw values byte-for-byte. No compiler spans the two
 * languages, so the contract is held by (a) this file staying visually
 * diff-able against its Swift twin, and (b) the mirrored test suite.
 * Change one side only in a commit that changes both.
 */
export const ROOM_STATES = [
  "LOBBY",
  "ACTIVE",
  "TALLY",
  "REVEALING",
  "MATCHED",
  "NO_RESULT",
] as const;

export type RoomState = (typeof ROOM_STATES)[number];

/** Same transition table as RoomState.allowedTransitions in Swift. */
const TRANSITIONS: Record<RoomState, readonly RoomState[]> = {
  // Start the round, or host ends before anything happened.
  LOBBY: ["ACTIVE", "NO_RESULT"],
  // All present finished, or the inactivity timeout fired; host may end.
  ACTIVE: ["TALLY", "NO_RESULT"],
  // Reveal, or overtime re-run (fresh deck / eliminate-lowest), or host end.
  TALLY: ["REVEALING", "ACTIVE", "NO_RESULT"],
  REVEALING: ["MATCHED", "NO_RESULT"],
  MATCHED: [],
  NO_RESULT: [],
};

export function isRoomState(value: string): value is RoomState {
  return (ROOM_STATES as readonly string[]).includes(value);
}

export function isTerminal(state: RoomState): boolean {
  return TRANSITIONS[state].length === 0;
}

export function canTransition(from: RoomState, to: RoomState): boolean {
  return TRANSITIONS[from].includes(to);
}

/** Swipes are only accepted while ACTIVE (REVEALING is input-closed). */
export function acceptsSwipes(state: RoomState): boolean {
  return state === "ACTIVE";
}
